executor/runtime/docker/docker.go

```diff
@@ -1640,6 +1649,137 @@ func (r *DockerRuntime) Start(parentCtx context.Context, pod *v1.Pod) (string, *
+		case <-ctx.Done():
+			l.WithError(ctx.Err()).Debug("Stopping reporting container status metric")
+			return
+		case <-reportTicker.C:
+			metricLines := r.containerStatusMetrics()
+			for _, m := range metricLines {
+				if _, err := conn.Write([]byte(m)); err != nil {
+					l.WithError(err).Warnf("Error sending metric to spectatord, metric string: %s", m)
+				}
+			}
+		}
+	}
+}
+
+func connectToSpectatord(spectatordSocket string, retriesLeft int) (net.Conn, error) {
+	conn, err := net.Dial("unixgram", spectatordSocket)
+	if err != nil {
+		if retriesLeft--; retriesLeft > 0 {
+			time.Sleep(1 * time.Second)
+			return connectToSpectatord(spectatordSocket, retriesLeft)
+		}
+	}
+	return conn, err
+}
+
+func (r *DockerRuntime) containerStatusMetrics() []string {
+	l := log.WithField("taskID", r.c.TaskID())
+	var metricLines []string
+	// 1. Loop through all containers.
+	for _, c := range r.c.Pod().Status.ContainerStatuses {
+		// 2. Get container metadata and state.
+		metricTags := make(map[string]string)
+		metricTags["nf.container"] = c.Name
+		metricTags["nf.process"] = processNameForContainer(c.Name, r.c.Pod())
+		imageName, imageVersion, err := imageNameAndDigest(c.Image, c.Name)
+		if err != nil {
+			l.WithError(err).Warnf("Error getting image metric tags for container status metrics")
+		}
+		if imageTag := imageTagForContainer(c.Name, r.c); imageTag != "" && imageName != "" {
+			imageName = fmt.Sprintf("%s:%s", imageName, imageTag)
+		}
+		metricTags["titus.image.name"] = imageName
+		metricTags["titus.image.version"] = imageVersion
+		state := "unhealthy"
+		if c.State.Running != nil {
+			state = "healthy"
+		}
+		metricTags["titus.state"] = state
+		// 3. Create the metric line using the spectatord protocol.
+		// The metric is a gauge with a value of 1 and a TTL of 2m.
+		statusGauge := fmt.Sprintf("g,2:titus.containers.status,%s:1", spectatordTags(metricTags))
+		metricLines = append(metricLines, statusGauge)
+	}
+	return metricLines
+}
+
+func processNameForContainer(containerName string, pod *v1.Pod) string {
+	for _, c := range pod.Spec.Containers {
+		if c.Name != containerName {
+			continue
+		}
+		for _, env := range c.Env {
+			// It's up to the process owners to set the NETFLIX_PROCESS_NAME value,
+			// no default value should be set if it's not present.
+			if env.Name == "NETFLIX_PROCESS_NAME" {
+				return env.Value
+			}
+		}
+	}
+	return ""
+}
+
+func imageNameAndDigest(image, containerName string) (name, digest string, err error) {
+	ref, err := reference.Parse(image)
+	if err != nil {
+		return "", "", fmt.Errorf("error parsing docker image %q for container %q: %w", image, containerName, err)
+	}
+	if named, isNamed := ref.(reference.Named); isNamed {
+		name = reference.Path(named)
+	}
+	if digested, isDigested := ref.(reference.Digested); isDigested {
+		digest = digested.Digest().String()
+	}
+	return name, digest, nil
+}
+
+func imageTagForContainer(containerName string, c runtimeTypes.Container) string {
```

Comment:
Your function here is  titus-kube-common `GetImageTagForContainer`

Reply:
Oh nice, I can remove this function then, as long as the plan is to follow the same annotation pattern (`AnnotationKeyImageTagPrefix + cName`) for other containers. The todo for us is adding the expected annotations to the pod.